### ISSUE TITLE
Rounding bugs

### ISSUE THREAD

**Issue by danielvonmitschke** [woocommerce/woocommerce#27847]
As stated in [this](https://github.com/woocommerce/woocommerce/issues/27626#issuecomment-696124721) comment there still are rounding "errors" in certain constellations - especially when entering net prices for shipping costs with three digits.

Of course there is always the option to set decimals to more than 2, but this will also display prices with more than 2 decimals.

But what I don't understand is why woocommerce does round prices at the "internal level" at all.
I think this is clearly wrong and woocommerce should do no rounding for internal prices and calculations.

An example for this done clearly the wrong way would be here:
https://github.com/woocommerce/woocommerce/blob/788969d2bbfba8f7e177b83504e70e98bc1babdf/includes/abstracts/abstract-wc-order.php#L652
Why do we need to round this value (shipping tax). Makes no sense (and leads to the bug that I described in the linked comment).

So my proposal would be to remove all internal rounding and apply the decimals options only to prices displayed to the customer.

For the time being we fortunately can get around this by using these two hooks:

Set all rounding to 3 decimals
```
add_filter('wc_get_price_decimals', function () {
    return 3;
});
```

Display prices with only 2 decimals (basically overriding the "wc_get_price_decimals" hook for displayed prices)
```
add_filter('wc_price_args', function ($args) {
    $args['decimals'] = 2;

    return $args;
});
```

But, again, I think it would be more correct to not round internal prices at all.

Best,

daniel

**Comment by danielvonmitschke** [comment]
Addon: As PayPal does not like 3 digits amounts we also have to take care to the amounts in refund objects.
This can be done with this hook:

```
        add_action('woocommerce_create_refund', function ($refund, $args) {
            if (!$refund instanceof \WC_Order_Refund) {
                return;
            }

            // Be sure to use only 2 decimals in refunds (paypal does not like 3 or more decimals)
            $refund->set_amount(wc_format_decimal($refund->get_amount(), 2));
            $refund->set_total(wc_format_decimal($refund->get_total(), 2));
        }, 10, 2);
```

**Comment by jrick1229** [comment]
Seen in: 25591798-hc

**Comment by vedanshujain** [comment]
I am closing this issue because the bug in the scope of this issue was addressed in #25800. However, we are still rounding in some of the internal calculations, but that will be addressed in a separate PR after we added unit tests around those cases.
